### ISSUE TITLE
Is it possible to share the code of Deep feature selection

### ISSUE THREAD

**Issue by WYBupup** [maksym33/FeatureImportanceDL#7]
Thank you for your effort on the paper and excellent idea.

I am now trying to reproduce the experiment mentioned in the paper. I have finished the experiment of the proposed dual-net on three synthetic data set and MNIST subset dataset. Now, I am trying to reproduce the experiment with the use of Deep feature selection [https://www.liebertpub.com/doi/abs/10.1089/cmb.2015.0189](url) on the same dataset. 

I code the model by myself and use the hyper-parameter mentioned in the supplementary material(lambda1=0.01, lambda2=1, alpha1=0.0001, alpha2=0. and the neural network's architecture and etc), however, the result on XOR synthetic data is not that good, and the categorial accuracy on the test set is just 0.25 which means the model just assign the same label to every sample. And I found that all the weights of the on to one layer is nearly zero under this hyper-parameter setting.

And then, I conducted hyperparameter tuning by myself using 5-fold cross-validation, however, the result shows that the smaller the lambda1 is the better the model is....  The result is shown below, where x-axis is log(lambda1).

![fig1](https://user-images.githubusercontent.com/61353633/148227642-379d2302-8130-4d4b-a06b-a3b361d23e52.png)

Following is the code of one to one layer
```
class One2oneLayer(tf.keras.layers.Layer):
    def __init__(self, lambda1=0.01, lambda2=1):
        super(One2oneLayer, self).__init__()
        self.lambda1 = lambda1
        self.lambda2 = lambda2

    def build(self, input_shape):
        self.w = self.add_weight(shape=(input_shape[-1],), initializer="random_normal",
                                 trainable=True)

    def call(self, inputs, **kwargs):
        loss = self.lambda1 * (1 - self.lambda2) / 2 * tf.nn.l2_loss(self.w) + \
               self.lambda1 * self.lambda2 * tf.reduce_sum(tf.abs(self.w))
        self.add_loss(loss)
        return tf.multiply(inputs, self.w)
```

Following is the model generation code of the DFS model 
```
    def create_dense_model(self, input_shape, dense_arch, sparse_input_layer_num=1, dense_activation="sigmoid",
                           last_activation="linear"):
        self.sparse_input_layer_num = sparse_input_layer_num
        self.x_shape = input_shape
        self.y_shape = dense_arch[-1]
        input_data_layer = Input(shape=input_shape)
        # flatten
        x = Flatten()(input_data_layer)
        # create one to one layer
        for i in range(sparse_input_layer_num):
            x = One2oneLayer(self.lambda1, self.lambda2)(x)
        # dense layers
        for units in dense_arch[:-1]:
            x = Dense(units=units, activation=dense_activation,
                      kernel_regularizer=tf.keras.regularizers.L1L2(l2=self.alpha1 * (1 - self.alpha2) / 2,
                                                                    l1=self.alpha1 * self.alpha2))(x)
        # last layer
        x = Dense(units=dense_arch[-1], activation=last_activation,
                  kernel_regularizer=tf.keras.regularizers.L1L2(l2=self.alpha1 * (1 - self.alpha2) / 2,
                                                                l1=self.alpha1 * self.alpha2))(x)
        self.model = Model(inputs=[input_data_layer], outputs=x)
        print("DFS dense model built:")
        self.model.summary()
```

I have struggled with this problem for a long time, and I can't figure where the problem is. So I think if you could share the code of the Deep feature selection model. It will help me a lot to figure out where I mess up. Thanks a lot.

**Comment by WYBupup** [comment]
I think that one of the reasons may be gradient vanishing since the author of the original paper mentioned that their algorithm may perform worse for a large number of hidden layers. 

> However, as the number of hidden layers increases, this algorithm would deteriorate, because gradient information disperses or explodes in lower layers

Would you please give me some hints about the initialization value of the weight or if the batch norm strategy is used for the deep feature selection model during the experiment or something else?

sorry for bothering.

**Comment by maksym33** [comment]
Hi,
from what I remember the DFS results were higlhy volatile - sometimes it would not be even close to the optimal performance, which is why we had slightly lower performance for it - if it did not converge for **one** of the CV runs, it would ruin the average performance. 
I had a look at your code and it looks nearly identical to the implementation we developed (which includes ` initializer="random_normal"`).
I have not used any batchnorm layers as we tried to keep the networks (DFS or the one mentioned in the paper) as vanilla as possible.

**Comment by WYBupup** [comment]
Thanks for your reply and help. 

The result of my experiment was that all 5 cross-validation test sets' accuracy was 0.25 on the XOR data set. And all of their one-to-one layer's weight is nearly 0 under the hyperparameter (lambda1=0.01, lambda2=1, alpha1=0.0001, alpha2=0).... which is annoying.

And I thought there may be a mistake in the implementation in TensorFlow since I am kind of new to TensorFlow. So, I also implemented the DFS in PyTorch, however, the result is nearly the same... 

And I also tried to use [https://github.com/hyperopt/hyperopt](url) to do hyperparameter tuning by 5 folds for all lambda1, lambda2, alpha1, alpha2. And the result is that (lambda1=0.0008, lambda2=0.0, alpha1=0.0001, alpha2=0.000
 which seems there is no L1 penalty on the loss function.  And following is the plot of lambda1 vs weights for every feature, and it seems there is no sparsity as the lambda1 increases and it seems just the phenomenon of L2 norm.
![myplot](https://user-images.githubusercontent.com/61353633/148672207-042570a1-35f6-404b-8054-bbe55f37d0f4.png)
001)

I think that I am running out of options on finding the reason... Although it is very inappropriate,  Is it possible to share your code of the DFS model with me?

And I will close the problem since this is not the problem of the proposed model.

Thanks for your open source code and detailed supplementary material of the paper!
